flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/CopyOnWriteStateTable.java

```diff
@@ -1032,75 +1040,143 @@ public int sizeOfNamespace(Object namespace) {
 
 	// StateEntryIterator  ---------------------------------------------------------------------------------------------
 
+	@Override
+	public StateIteratorWithUpdate<K, N, S> getStateEntryIterator() {
+		return new StateEntryIterator(true);
+	}
+
 	/**
-	 * Iterator over the entries in a {@link CopyOnWriteStateTable}.
+	 * Iterator over state entries in a {@link CopyOnWriteStateTable}.
+	 *
+	 * <p>This iterator can optionally tolerate concurrent modifications and
+	 * support removing of underlying next key state.
 	 */
-	class StateEntryIterator implements Iterator<StateEntry<K, N, S>> {
+	class StateEntryIterator implements StateIteratorWithUpdate<K, N, S> {
```

Comment:
I would suggest not to mix two iterators into one implementation and deceide behaviour on a boolean flag. Better have a common superclass and replace the boolean flag by polymorphic methods. Then you can also only have the subclass "with update" implement the full `StateIteratorWithUpdate` whereas the other can still simply work as `Iterator`